crates/fluvio-spu/src/kv/consumer.rs

```diff
@@ -0,0 +1,295 @@
+        Self {
+            replica_id: replica_id.into(),
+            consumer_id: consumer_id.into(),
+        }
+    }
+}
+impl Consumer {
+    pub(crate) fn new(offset: i64, ttl: Duration) -> Self {
+        let now = now_timestamp();
+        Self::with(offset, ttl, now + ttl.as_secs())
+    }
+
+    pub(crate) fn with(offset: i64, ttl: Duration, expire_time: TimestampSecs) -> Self {
+        Self {
+            offset,
+            ttl,
+            expire_time,
+        }
+    }
+}
+
+impl From<ConsumerStorage> for SharableConsumerStorage {
+    fn from(value: ConsumerStorage) -> Self {
+        Self(Arc::new(RwLock::new(value)))
+    }
+}
+
+impl SharableConsumerStorage {
+    pub async fn get_by(
+        &self,
+        topic: String,
+        partition: PartitionId,
+        consumer_id: String,
+    ) -> Result<Option<Consumer>> {
+        let replica: ReplicaKey = (topic, partition).into();
+        let key = Key::new(replica, consumer_id);
+        self.get(&key).await
+    }
+
+    pub async fn get(&self, key: &Key) -> Result<Option<Consumer>> {
+        self.0.read().await.get(key).await
+    }
+
+    pub async fn delete(&self, key: &Key) -> Result<()> {
+        self.0.write().await.delete(key).await
+    }
+
+    pub async fn put(&self, key: impl Into<Key>, value: impl Into<Consumer>) -> Result<()> {
+        self.0.write().await.put(key, value).await
+    }
+}
+
+fn now_timestamp() -> TimestampSecs {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap_or_default()
+        .as_secs()
+}
+
+fn is_not_expired(record: &Consumer) -> bool {
+    let now = now_timestamp();
+    now <= record.expire_time
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{env::temp_dir, path::Path};
+
+    use fluvio_controlplane::replica::Replica;
+    use fluvio_storage::config::ReplicaConfig;
+    use fluvio_types::PartitionId;
+    use flv_util::fixture::ensure_clean_dir;
+
+    use crate::{
+        config::ReplicationConfig, storage::SharableReplicaStorage,
+        control_plane::StatusMessageSink,
+    };
+
+    use super::*;
+
+    #[fluvio_future::test]
+    async fn test_expiration_checked() {
+        //given
+        let leader = create_replica("test_expiration_checked").await;
+        let notifier = FollowerNotifier::shared();
+        let mut storage = ConsumerStorage::new(leader.clone(), notifier);
+        let key1 = Key::new(("topic1", 0), "consumer1");
+        let value1 = Consumer::with(1, Duration::from_secs(2), now_timestamp() - 100);
+        let key2 = Key::new(("topic2", 1), "consumer1");
+        let value2 = Consumer::with(2, Duration::from_secs(2), now_timestamp() + 100);
+        storage
+            .put(key1.clone(), value1)
+            .await
+            .expect("put record1");
+        storage
+            .put(key2.clone(), value2.clone())
+            .await
+            .expect("put record2");
+
+        //when
+        let record1 = storage.get(&key1).await.expect("record1");
```

Comment:
confused. is this getting record or offset?

Reply:
It is getting a value from the storage. The type of the value is `ConsumerOffset`, `record1` is a variable name

Reply:
Still confused about need to have TTL in the offset value.  It doesn't make sense.   We should provide minimum API, which in case just need incremental offset value.   
In either case, this test case is misleading because there is no testing of TTL anyway.   So either test should case reflect TTL or provide API just with offset so only need to check if offset is persist

Reply:
Again, TTL is needed to show on CLI. This test does not test TTl, it test that the storage does not return expired values. Expired means `expiration_time` < `now`